pubsub/azure/servicebus/topics/servicebus.go

```diff
@@ -298,3 +315,146 @@ func (a *azureServiceBus) Close() (err error) {
+	receiver, err := sub.Connect(func() (impl.Receiver, error) {
+		a.logger.Debugf("Connecting to subscription %s for topic %s", a.metadata.ConsumerID, req.Topic)
+		r, err := a.client.GetClient().NewReceiverForSubscription(req.Topic, a.metadata.ConsumerID, nil)
+		return impl.NewMessageReceiver(r), err
+	})
+	if err != nil {
+		// Realistically, the only time we should get to this point is if the context was canceled, but let's log any other error we may get.
+		if !errors.Is(err, context.Canceled) {
+			a.logger.Errorf("Could not instantiate session subscription %s to topic %s", a.metadata.ConsumerID, req.Topic)
+		}
+		return nil
+	}
+	defer func() {
+		closeReceiverCtx, closeReceiverCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(a.metadata.TimeoutInSec))
+		receiver.Close(closeReceiverCtx)
+		closeReceiverCancel()
+	}()
+
+	// lock renewal loop
+	go func() {
+		a.logger.Debugf("Renewing locks for subscription %s for topic %s", a.metadata.ConsumerID, req.Topic)
+		lockErr := sub.RenewLocksBlocking(receiverCtx, receiver, impl.LockRenewalOptions{
+			RenewalInSec: a.metadata.LockRenewalInSec,
+			TimeoutInSec: a.metadata.TimeoutInSec,
+		})
+		if lockErr != nil {
+			a.logger.Error(lockErr)
+		}
+	}()
+
+	a.logger.Debugf("Receiving messages from subscription %s for topic %s", a.metadata.ConsumerID, req.Topic)
+
+	// receiveAndBlockFn will only return with an error that it cannot handle internally. The subscription connection is closed when this method returns.
+	// If that occurs, we will log the error and attempt to re-establish the subscription connection until we exhaust the number of reconnect attempts.
+	if err := receiveAndBlockFn(receiver, onFirstSuccess); err != nil {
+		return err
+	}
+
+	// Gracefully close the connection (in case it's not closed already)
+	// Use a background context here (with timeout) because ctx may be closed already.
+	closeCtx, closeCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(a.metadata.TimeoutInSec))
+	sub.Close(closeCtx)
+	closeCancel()
+
+	return nil
+}
+
+func (a *azureServiceBus) ConnectAndReceiveWithSessions(subscribeCtx context.Context, req pubsub.SubscribeRequest, sub *impl.Subscription, receiveAndBlockFn func(impl.Receiver, func()) error, onFirstSuccess func(), maxConcurrentSessions int) {
+	sessionsChan := make(chan struct{}, maxConcurrentSessions)
+	for i := 0; i < maxConcurrentSessions; i++ {
+		sessionsChan <- struct{}{}
+	}
+
+	defer func() {
+		// Gracefully close the connection (in case it's not closed already)
+		// Use a background context here (with timeout) because ctx may be closed already
+		closeSubCtx, closeSubCancel := context.WithTimeout(context.Background(), time.Second*time.Duration(a.metadata.TimeoutInSec))
+		sub.Close(closeSubCtx)
+		closeSubCancel()
+	}()
+
+	for {
+		select {
+		case <-subscribeCtx.Done():
+			return
+		case <-sessionsChan:
+			select {
+			case <-subscribeCtx.Done():
```

Comment:
Here too, why is this repeating the `select` on the same case?

Reply:
see above